### PR TITLE
[openshift-resources] add jumphost

### PR DESCRIPTION
**Pull request by maorfr**

### CODE REVIEW

File: utils/oc.py [hunk 1]
```diff
@@ -15,12 +16,30 @@ class JSONParsingError(Exception):
...
+
+        import tempfile
+
+        hostname = jh_data['hostname']
+        identity = jh_data['identity']
+        user = jh_data['user']
+        identity_file = tempfile.mkdtemp() + '/id'
+        with open(identity_file, 'w') as f:
+            f.write(identity)
+        s.chmod(identity_file, 0o600)
```
Comment (riuvshin) [inline utils/oc.py]: this looks weird

Comment (maorfr) [reply utils/oc.py]: this is octal representation

Comment (riuvshin) [reply utils/oc.py]: wasn't aware that 0600 will not work with Py3, looks good then.

File: reconcile/openshift_resources.py [hunk 2]
```diff
@@ -466,13 +480,19 @@ def realize_data(dry_run, oc_map, ri):
...
+        oc.cleanup()
+
+
 def run(dry_run=False, thread_pool_size=10):
     gqlapi = gql.get_api()
 
     namespaces_query = gqlapi.query(NAMESPACES_QUERY)['namespaces']
 
     oc_map, ri = fetch_data(namespaces_query, thread_pool_size)
     realize_data(dry_run, oc_map, ri)
```
Comment (jmelis) [inline reconcile/openshift_resources.py]: can you change this to:

```python
try:
    oc_map, ri = fetch_data(namespaces_query, thread_pool_size)
    realize_data(dry_run, oc_map, ri)
finally:
    cleanup(oc_map)
```

Comment (maorfr) [reply reconcile/openshift_resources.py]: good idea!

File: utils/oc.py [hunk 3]
```diff
@@ -15,12 +17,25 @@ class JSONParsingError(Exception):
...
+            oc_base_cmd = jump_host.ssh_base_cmd + oc_base_cmd
+            self.jump_host = jump_host
+
+        self.oc_base_cmd = oc_base_cmd
 
     def whoami(self):
         return self._run(['whoami'])
 
+    def cleanup(self):
+        if not hasattr(self, 'jump_host'):
```
Comment (jmelis) [inline utils/oc.py]: `if not hasattr(self, 'jump_host') and isinstance(self.jump_host, JumpHost)`

File: utils/jump_host.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import utils.vault_client as vault_client
+
+
+class JumpHost(object):
+    def __init__(self, jh):
+        self.hostname = jh['hostname']
+        self.user = jh['user']
+        self.port = '22' if jh['port'] is None else jh['port']
```
[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: jf :)

[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: this is much cleaner, but will not work as expected, as jh['port'] is either a number or `None`, and self.port gets assigned `None`.

[the remark this answers is not in the document]
Comment (jmelis) [reply utils/jump_host.py]: you are totally right, sorry about that!

File: utils/oc.py (hunk 3 above)
Comment (jmelis) [reply utils/oc.py]: (what I want to do is to prevent that for some reason jump_host is None or something)

Comment (maorfr) [reply utils/oc.py]: fixed (`or not` instead of `and`)

File: utils/jump_host.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        with open(identity_file, 'w') as f:
+            f.write(self.identity)
+        os.chmod(identity_file, 0o600)
+        user_host = '{}@{}'.format(self.user, self.hostname)
+
+        self.ssh_base_cmd = \
+            ['ssh', '-i', identity_file, '-p', self.port, user_host]
+
+    def cleanup(self):
+        import shutil
```
[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: fixed

File: utils/jump_host.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        import os
+
+        self._identity_dir = tempfile.mkdtemp()
+        identity_file = self._identity_dir + '/id'
+        with open(identity_file, 'w') as f:
+            f.write(self.identity)
+        os.chmod(identity_file, 0o600)
+        user_host = '{}@{}'.format(self.user, self.hostname)
+
+        self.ssh_base_cmd = \
```
[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: added a setter and a getter. is this what you had in mind?

File: utils/jump_host.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            vault_client.read(jh_identity['path'], jh_identity['field'])
+        if jh_identity['format'] == 'base64':
+            import base64
+
+            identity = base64.b64decode(identity)
+        return identity
+
+    def init_ssh_base_cmd(self):
+        import tempfile
+        import os
```
[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: fixed

File: utils/jump_host.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        self.identity = self.get_identity_from_vault(jh)
+
+        self.init_ssh_base_cmd()
+
+    def get_identity_from_vault(self, jh):
+        jh_identity = jh['identity']
+        identity = \
+            vault_client.read(jh_identity['path'], jh_identity['field'])
+        if jh_identity['format'] == 'base64':
+            import base64
```
[the remark this answers is not in the document]
Comment (maorfr) [reply utils/jump_host.py]: fixed

**Pull request merged**